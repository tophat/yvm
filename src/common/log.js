const defaultLogger = console.error.bind(console) // eslint-disable-line no-console
const capturableLogger = console.log.bind(console) // eslint-disable-line no-console

function log(...args) {
    defaultLogger(...args)
}

log.capturable = function capturableLog(...args) {
    capturableLogger(...args)
}

module.exports = log
