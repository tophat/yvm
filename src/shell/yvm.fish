#!/usr/bin/env fish
set -q YVM_DIR; or set -gx YVM_DIR "$HOME/.yvm"

function yvm
    set command $argv[1]

    function set_fish_user_paths
        set -gx fish_user_paths (string split ' ' -- $argv[1])
    end

    function yvm_use
        if test (count $argv) -gt 0
            set PROVIDED_VERSION $argv[1]
        end

        begin
            set NEW_FISH_USER_PATHS (yvm_call_node_script get-new-path --shell=fish $PROVIDED_VERSION)
        end
        if [ -z "$NEW_FISH_USER_PATHS" ]
            yvm_err "Could not get new path from yvm"
            exit 1
        else
            set_fish_user_paths $NEW_FISH_USER_PATHS
            set -l new_version (yarn --version)
            yvm_echo "Now using yarn version $new_version"
        end
    end

    function yvm_shim
        begin
            set NEW_FISH_USER_PATHS (yvm_call_node_script get-shim-path --shell=fish)
        end
        if [ -z "$NEW_FISH_USER_PATHS" ]
            yvm_err "Could not get shim path from yvm"
            exit 1
        else
            set_fish_user_paths $NEW_FISH_USER_PATHS
        end
    end

    function yvm_deactivate
        begin
            set NEW_FISH_USER_PATHS (yvm_call_node_script get-old-path --shell=fish)
        end
        if [ -z "$NEW_FISH_USER_PATHS" ]
            yvm_err "Could not remove yvm from system path"
            exit 1
        else
            set_fish_user_paths $NEW_FISH_USER_PATHS
        end
    end

    function yvm_unload
        yvm_deactivate
        set -e YVM_DIR
        functions -e set_fish_user_paths
        functions -e yvm_use
        functions -e yvm_shim
        functions -e yvm_deactivate
        functions -e yvm_unload
        functions -e yvm_err
        functions -e yvm_call_node_script
        functions -e yvm_init_sh
        functions -e yvm
        yvm_echo "YVM configuration unloaded from shell"
        functions -e yvm_echo
    end

    function yvm_echo
        command printf "%s\n" $argv 2>/dev/null
    end

    function yvm_err
        yvm_echo "$argv" >&2
    end

    function yvm_call_node_script
        set -lx fish_user_paths $fish_user_paths
        node "$YVM_DIR/yvm.js" $argv
    end

    function yvm_init_sh
        if not type -q "node"
            yvm_err "%s\n" "YVM Could not automatically set yarn version."
            yvm_err "%s\n" "Please ensure your YVM env variables and sourcing are set below sourcing node/nvm in your fish config file"
            exit 1
        end
        yvm_shim
    end

    if [ "$command" = "use" ]
        if test (count $argv) -gt 1
            yvm_use $argv[2]
        else
            yvm_use
        end
    else if [ "$command" = "update-self" ]
        env YVM_INSTALL_DIR=$YVM_DIR curl -fsSL https://raw.githubusercontent.com/tophat/yvm/master/scripts/install.js | node
    else if [ "$command" = "init-sh" ]
        yvm_init_sh
    else if [ "$command" = "shim" ]
        yvm_shim
        yvm_echo "Now shimming yarn"
    else if [ "$command" = "deactivate" ]
        yvm_deactivate
        yvm_echo "YVM shim and yarn versions removed from PATH"
    else if [ "$command" = "unload" ]
        yvm_unload
    else
        yvm_call_node_script $argv[1..-1]
    end
end

if count $argv >/dev/null
    yvm $argv
else
    yvm init-sh
end
