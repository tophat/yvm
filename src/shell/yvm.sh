#!/bin/sh
YVM_DIR=${YVM_DIR-"${HOME}/.yvm"}

yvm() {
    command=$1

    yvm_use() {
        local PROVIDED_VERSION=${1}
        NEW_PATH=$(yvm_call_node_script get-new-path ${PROVIDED_VERSION})
        if [ -z "${NEW_PATH}" ]; then
            yvm_err "Could not get new path from yvm"
        else
            PATH=${NEW_PATH}
            yvm_echo "Now using yarn version $(yarn --version)"
        fi
    }

    yvm_echo() {
        command printf %s\\n "$*" 2>/dev/null
    }

    yvm_err() {
        >&2 yvm_echo "$@"
    }

    yvm_call_node_script() {
        # do not add anything that outputs stuff to stdout in function, its output is stored in a variable
        node "${YVM_DIR}/yvm.js" $@
    }

    yvm_init_sh() {
        if ! type "node" > /dev/null; then
            yvm_err "YVM Could not find node executable."
            yvm_err "Please ensure your YVM env variables and sourcing are set below sourcing node/nvm in your .zshrc or .bashrc"
        fi
        export PATH="${YVM_DIR}/shim":$PATH
    }

    if [ "${command}" = "use" ]; then
        yvm_use $2
    elif [ "${command}" = "update-self" ]; then
        curl -fsSL https://raw.githubusercontent.com/tophat/yvm/master/scripts/install.js | YVM_INSTALL_DIR=${YVM_DIR} node
    elif [ "${command}" = "init-sh" ]; then
        yvm_init_sh
    else
        yvm_call_node_script $@
    fi
}

if [ -n "$*" ]; then
    yvm $@
else
    yvm init-sh
fi
