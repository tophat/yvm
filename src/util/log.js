const chalk = require('chalk')

const defaultLogger = console.error.bind(console) // eslint-disable-line no-console
const capturableLogger = console.log.bind(console) // eslint-disable-line no-console

const log = (...args) => log.default(...args)

log.default = function defaultLog(...args) {
    defaultLogger(...args)
}

log.capturable = function capturableLog(...args) {
    capturableLogger(...args)
}

log.info = function infoLog(...args) {
    if (process.env.YVM_VERBOSE) {
        log(...args)
    }
}

log.error = function errorLog(...args) {
    log(chalk.red(...args))
}

log.success = function successLog(...args) {
    log(chalk.green(...args))
}

log.notice = function noticeLog(...args) {
    log(chalk.yellow(...args))
}

module.exports = log
