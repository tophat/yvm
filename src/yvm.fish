#!/bin/sh
set -q YVM_DIR; or set -gx YVM_DIR "$HOME/.yvm"

function yvm
    set command $argv[1]

    function yvm_use
        if test (count $argv) -gt 0
            set PROVIDED_VERSION $argv[1]
        end

        begin
            set NEW_FISH_USER_PATHS (yvm_call_node_script get-new-path --shell=fish $PROVIDED_VERSION)
        end
        if [ -z "$NEW_FISH_USER_PATHS" ]
            yvm_err "Could not get new path from yvm"
        else
            set -gx fish_user_paths (string split ' ' -- $NEW_FISH_USER_PATHS)
            set -l new_version (yarn --version)
            yvm_echo "Now using yarn version $new_version"
        end
    end

    function yvm_echo
        command printf "%s\n" $argv 2>/dev/null
    end

    function yvm_err
        yvm_echo "$argv" >&2
    end

    function yvm_call_node_script
        set -lx fish_user_paths $fish_user_paths
        node "$YVM_DIR/yvm.js" $argv
    end

    function yvm_init_sh
        if not type -q "node"
            yvm_err "%s\n" "YVM Could not automatically set yarn version."
            yvm_err "%s\n" "Please ensure your YVM env variables and sourcing are set below sourcing node/nvm in your fish config file"
            exit 1
        end
        set -gx fish_user_paths "$YVM_DIR/shim" $fish_user_paths
    end

    if [ "$command" = "use" ]
        if test (count $argv) -gt 1
            yvm_use $argv[2]
        else
            yvm_use
        end
    else if [ "$command" = "update-self" ]
        env YVM_INSTALL_DIR=$YVM_DIR curl -fsSL https://raw.githubusercontent.com/tophat/yvm/master/scripts/install.js | node
    else if [ "$command" = "init-sh" ]
        yvm_init_sh
    else
        yvm_call_node_script $argv[1..-1]
    end
end

yvm init-sh
