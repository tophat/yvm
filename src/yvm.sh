#!/bin/sh

command=$1
YVM_DIR=${YVM_DIR-"${HOME}/.yvm"}

yvm_use() {
    local PROVIDED_VERSION=${1}
    NEW_PATH=$(yvm_call_node_script get-new-path ${PROVIDED_VERSION})
    if [ -z "${NEW_PATH}" ]; then
        yvm_err "Could not get new path from yvm"
    else
        PATH=${NEW_PATH}
    fi
}

yvm_echo() {
    command printf %s\\n "$*" 2>/dev/null
}

yvm_err() {
    >&2 yvm_echo "$@"
}

yvm_call_node_script() {
    # do not add anything that outputs stuff to stdout in function, its output is stored in a variable
    node "${YVM_DIR}/yvm.js" $@
}

case "$-" in
    *i*) interactive=1;;
    *) interactive=0;;
esac

yvm_() {
    mode=$1
    shift 1

    command=$1
    if [ "${command}" = "use" ]; then
        if [ "${mode}" = 'script' ]; then
            yvm_err '"yvm use" can only be used when yvm is a shell function, not a script. Did you forget to source yvm?'
            exit 1
        fi
        yvm_use $2
    elif [ "${command}" = "update-self" ]; then
        curl -fsSL https://raw.githubusercontent.com/tophatmonocle/yvm/master/scripts/install.sh | YVM_INSTALL_DIR=${YVM_DIR} bash
    else
        yvm_call_node_script $@
    fi
}

if [ ${interactive} = 1 ]; then
    yvm() {
        yvm_ 'function' $@
    }
else
    yvm_ 'script' $@
fi
