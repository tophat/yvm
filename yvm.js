const argParser = require('commander')

console.log('called yvm.js with args', process.argv)

argParser
    .command('install <version>')
    .action(version => {
        console.log(`Installing yarn v${version}`)
    })

argParser
    .command('remove <version>')
    .action(version => {
        console.log(`Removing yarn v${version}`)
    })

argParser
    .command('manage <version>')
    .action(version => {
        console.log(`Managing yarn v${version}`)
    })

argParser.parse(process.argv)
