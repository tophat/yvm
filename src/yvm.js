import argParser from 'commander'

import log from 'util/log'
import { yvmInstalledVersion } from 'util/yvmInstalledVersion'

const messageNoYvm = 'Unable to determine yvm version'
const messageSourceYvm =
    'You need to source yvm to use this command. run `source ~/.yvm/yvm.sh`'
const messageOptionalVersion = msg => `${msg}. Uses config file if available.`
const signPosting = command => `See -> 'yvm ${command}'`

argParser.description('Yarn Version Manager')
argParser.addImplicitHelpCommand = () => {}

if (!process.argv.includes('exec')) {
    argParser.version(yvmInstalledVersion() || messageNoYvm)
    argParser
        .command('*', '', { noHelp: true, isDefault: true })
        .action(invalidCommand => {
            if (!process.argv.includes('help')) {
                log(`Invalid command: ${invalidCommand}`)
            }
            argParser.outputHelp()
            process.exit(1)
        })
}

argParser
    .command('install [version]')
    .alias('i')
    .option('-l, --latest', signPosting`install latest`)
    .option('-s, --stable', signPosting`install stable`)
    .option('--verify', 'Verifies GPG signature')
    .description(messageOptionalVersion`Install the specified version of Yarn`)
    .action(async (maybeVersion, command) => {
        const { install } = await import('commands/install')
        const exitCode = await install({
            latest: command.latest,
            stable: command.stable,
            version: maybeVersion,
            verifyGPG: command.verify,
        })
        process.exit(exitCode)
    })

const uninstallVersion = async version => {
    const { remove } = await import('commands/remove')
    process.exit(await remove(version))
}
argParser
    .command('uninstall <version>')
    .alias('rm')
    .description('Uninstall the specified version of Yarn.')
    .action(uninstallVersion)

argParser
    .command('use [version]')
    .description(messageOptionalVersion`Activate specified Yarn version`)
    .action(() => log(messageSourceYvm))

argParser
    .command('exec [version] [args...]')
    .option('-h, --help')
    .allowUnknownOption()
    .description(
        messageOptionalVersion`Execute command using specified Yarn version`,
    )
    .action(async () => {
        const [, , , maybeVersion, ...rest] = process.argv
        const { exec } = await import('commands/exec')
        process.exit(await exec(maybeVersion, rest))
    })

argParser
    .command('current')
    .description('Display current active Yarn version')
    .action(async command => {
        log.info('Checking Yarn version')
        const { current } = await import('commands/current')
        process.exit(await current(command))
    })

argParser
    .command('list')
    .alias('ls')
    .description('List the currently installed versions of Yarn.')
    .action(async () => {
        const { list } = await import('commands/list')
        process.exit(await list())
    })

argParser
    .command('list-remote')
    .alias('ls-remote')
    .description('List Yarn versions available to install.')
    .action(async () => {
        const { listRemote } = await import('commands/listRemote')
        process.exit(await listRemote())
    })

argParser.command('alias [<pattern>]', 'Show all aliases matching <pattern>')
argParser.command(
    'alias <name> <version>',
    'Set an alias named <name> pointing to <version>',
)
argParser.command('alias', '', { noHelp: true }).action(async () => {
    const [, , , nameOrPattern, maybeVersion] = process.argv
    const { alias } = await import('commands/alias')
    process.exit(await alias(nameOrPattern, maybeVersion))
})

argParser
    .command('unalias <name>')
    .option('-F, --force', 'Delete even if other aliases will be broken')
    .option('-R, --recursive', 'Delete all dependant aliases as well')
    .description('Deletes the alias named <name>')
    .action(async (name, { force, recursive }) => {
        const { unalias } = await import('commands/unalias')
        process.exit(await unalias({ name, force, recursive }))
    })

argParser
    .command('which [version]')
    .description(messageOptionalVersion`Display path to Yarn version`)
    .action(async maybeVersion => {
        log.info('Getting path to Yarn version')
        const { which } = await import('commands/which')
        process.exit(await which({ version: maybeVersion }))
    })

argParser
    .command('update-self')
    .description('Updates yvm to the latest version')
    .action(() => log(messageSourceYvm))

argParser
    .command('configure-shell', '', { noHelp: true })
    .option('--home [home]', 'Alternate path to home directory')
    .option('--shell [shell]', 'Shell to configure, defaults to all')
    .option('--no-shim', 'Skips yarn shim')
    .description(
        'Internal command: Configures any shell config files found for loading yvm on startup',
    )
    .action(async ({ home, shell, shim }) => {
        const { configureShell } = await import('commands/configureShell')
        process.exit(await configureShell({ home, shell, shim }))
    })

argParser
    .command('get-new-path [version]', '', { noHelp: true })
    .option('--shell [shell]', 'Shell used when getting PATH')
    .description(
        'Internal command: Gets a new PATH string for "yvm use", installing the version if necessary',
    )
    .action(async (maybeVersion, { shell }) => {
        const { getNewPath } = await import('commands/getNewPath')
        process.exit(await getNewPath(maybeVersion, shell))
    })

argParser
    .command('version')
    .description(signPosting`--version`)
    .action(() => {
        const version = yvmInstalledVersion()
        log.capturable(version || messageNoYvm)
        process.exit(version ? 0 : 1)
    })

argParser
    .command('remove <version>')
    .description(signPosting`uninstall`)
    .action(uninstallVersion)

argParser
    .command('set-default <version>')
    .description(signPosting`alias default <version>`)
    .action(async version => {
        const { setDefault } = await import('commands/setDefault')
        process.exit(await setDefault(version))
    })

argParser
    .command('get-default-version')
    .description(signPosting`alias default`)
    .action(async () => {
        const { getDefaultVersion } = await import('commands/getDefaultVersion')
        process.exit(await getDefaultVersion())
    })

argParser.parse(process.argv)
