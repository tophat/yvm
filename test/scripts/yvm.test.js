import mockProps from 'jest-mock-props'
import log from 'util/log'
import { program } from 'yvm'

mockProps.extend(jest)

describe('yvm commands', () => {
    jest.spyOn(process, 'exit').mockImplementation(jest.fn())
    const argSpy = jest.spyOnProp(process, 'argv')
    const stdOutSpy = jest.fn()
    const stdErrSpy = jest.fn()
    const mockStdOut = (...args) =>
        jest.spyOn(...args).mockImplementation(s => stdOutSpy(String(s)))
    const mockStdErr = (...args) =>
        jest.spyOn(...args).mockImplementation(s => stdErrSpy(String(s)))

    const concatMockCalls = spy => {
        return spy.mock.calls
            .map(args => args.join(' ').trim())
            .join('\n')
            .trim()
    }

    const flush = () => new Promise(resolve => setTimeout(resolve))

    const runWithArgs = async (...args) => {
        argSpy.mockValue(['node', 'yvm.js', ...args])
        program(process.argv).parse(process.argv)
        await flush()
        return {
            stderr: concatMockCalls(stdErrSpy),
            stdout: concatMockCalls(stdOutSpy),
        }
    }

    beforeAll(() => {
        mockStdOut(process.stdout, 'write')
        mockStdOut(console, 'log')
        mockStdOut(log, 'capturable')
        mockStdErr(process.stderr, 'write')
        mockStdErr(console, 'error')
        mockStdErr(log, 'default')
    })
    beforeEach(jest.clearAllMocks)
    afterAll(jest.restoreAllMocks)

    describe('default', () => {
        it('shows error message when wrong args passed', async () => {
            expect((await runWithArgs('wrong', 'args')).stderr).toContain(
                'Invalid command: wrong',
            )
            expect(process.exit).toHaveBeenCalledWith(1)
        })
    })

    describe('help', () => {
        it('shows all help with --help', async () => {
            expect((await runWithArgs('--help')).stdout).toMatchSnapshot()
            expect(process.exit).toHaveBeenCalledWith(0)
        })
    })

    describe('use', () => {
        it('does not use in non shell context', async () => {
            expect((await runWithArgs('use')).stderr).toContain(
                'You need to source yvm to use this command. run `source ~/.yvm/yvm.sh`',
            )
            expect(process.exit).toHaveBeenCalledWith(1)
        })
    })

    describe('current', () => {
        it('shows yarn version', async () => {
            expect((await runWithArgs('current')).stdout).toMatchSnapshot()
            expect(process.exit).toHaveBeenCalledWith(0)
        })
    })
})
