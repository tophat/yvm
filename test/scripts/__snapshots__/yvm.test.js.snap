// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yvm commands current shows yarn version 1`] = `"1.13.0"`;

exports[`yvm commands help shows all help with --help 1`] = `
"Usage: yvm [options] [command]

Yarn Version Manager

Options:
  -V, --version                       output the version number
  -h, --help                          output usage information

Commands:
  install|i [options] [version]       Install the specified version of Yarn. Uses config file if available.
  uninstall|rm <version>              Uninstall the specified version of Yarn.
  use [version]                       Activate specified Yarn version. Uses config file if available.
  exec [options] [version] [args...]  Execute command using specified Yarn version. Uses config file if available.
  current                             Display current active Yarn version
  list|ls                             List the currently installed versions of Yarn.
  list-remote|ls-remote               List Yarn versions available to install.
  alias [<pattern>]                   Show all aliases matching <pattern>
  alias <name> <version>              Set an alias named <name> pointing to <version>
  unalias [options] <name>            Deletes the alias named <name>
  which [version]                     Display path to Yarn version. Uses config file if available.
  update-self                         Updates yvm to the latest version
  version                             See -> 'yvm --version'
  remove <version>                    See -> 'yvm uninstall'
  set-default <version>               See -> 'yvm alias default <version>'
  get-default-version                 See -> 'yvm alias default'"
`;
